src/main/java/org/broadinstitute/hellbender/tools/funcotator/FuncotatorDataSourceDownloader.java

```diff
@@ -0,0 +1,326 @@
+            optional = true)
+    private boolean getSomaticDataSources = false;
+
+    @Argument(fullName = GERMLINE_ARG_LONG_NAME,
+            shortName = GERMLINE_ARG_LONG_NAME,
+            mutex = {SOMATIC_ARG_LONG_NAME},
+            doc = "Download the latest pre-packaged datasources for germline functional annotation.",
+            optional = true)
+    private boolean getGermlineDataSources = false;
+
+    @Argument(fullName = OVERWRITE_ARG_LONG_NAME,
+            shortName = OVERWRITE_ARG_LONG_NAME,
+            doc = "Overwrite output file if it exists already.",
+            optional = true)
+    private boolean overwriteOutputFile = false;
+
+    @Argument(
+            shortName = StandardArgumentDefinitions.OUTPUT_SHORT_NAME,
+            fullName  = StandardArgumentDefinitions.OUTPUT_LONG_NAME,
+            doc = "Output location for the data sources.",
+            optional = true)
+    protected File outputFile;
+
+    @Argument(
+            shortName = EXTRACT_AFTER_DOWNLOAD,
+            fullName  = EXTRACT_AFTER_DOWNLOAD,
+            doc = "Extract the data sources to a sibling folder after they have been downloaded.",
+            optional = true)
+    protected boolean extractDataSourcesAfterDownload = false;
+
+    //==================================================================================================================
+    // Constructors:
+
+    //==================================================================================================================
+    // Override Methods:
+
+    @Override
+    protected void onStartup() {
+        // Make sure the user specified at least one data source to download:
+        if ((!getSomaticDataSources) && (!getGermlineDataSources)) {
+            throw new UserException("Must select either somatic or germline datasources.");
+        }
+
+        if ( overwriteOutputFile ) {
+            logger.info("Overwrite ENABLED.  Will overwrite existing data sources download.");
+        }
+    }
+
+    @Override
+    protected Object doWork() {
+
+        // Get the correct data source:
+        if ( getSomaticDataSources ) {
+            downloadAndValidateDataSources("Somatic", SOMATIC_GCLOUD_DATASOURCES_PATH, SOMATIC_GCLOUD_DATASOURCES_SHA256_PATH);
+        }
+
+        if ( getGermlineDataSources ) {
+            downloadAndValidateDataSources("Germline", GERMLINE_GCLOUD_DATASOURCES_PATH, GERMLINE_GCLOUD_DATASOURCES_SHA256_PATH);
+        }
+
+        // Token return value:
+        return true;
+    }
+
+    //==================================================================================================================
+    // Static Methods:
+
+    //==================================================================================================================
+    // Instance Methods:
+
+    private void extractDataSources(final Path localDataSourcesPath) {
```

Comment:
Can you add a good unit test for `extractDataSource()` / `extractFilesFromArchiveStream()`? You might also want to move `extractFilesFromArchiveStream()` to `IOUtils` rather than embed it in this tool.

Alternatively, if you can find a library method in our dependencies (apache, guava, etc.) that does the entire extraction process in a single method call, unit tests won't be necessary.

Reply:
I'll move it to IOUtils.

The code I have there is what Apache recommends for doing extraction and leverages some of their classes.  I based it on the information they have here: https://commons.apache.org/proper/commons-compress/examples.html

I'll write a unit test for it.